### ISSUE TITLE
Make removing Linkerd easier

### ISSUE THREAD

**Issue by grampelberg** [linkerd/linkerd2#2376]
### What problem are you trying to solve?

I'm writing up uninstall instructions and it would be nice to have a single `kubectl` command that selects everything with a proxy to uninject. Right now, it appears that `linkerd.io/control-plane-ns` is the only common label to select on (and that's not on the control plane components).

### How should the problem be solved?

Add a label `linkerd.io/injected: true` to all parent resources that have been injected.

### Any alternatives you've considered?

- Annotation instead - can't be used as a selector with `kubectl`.
- Label on the pod spec instead of the parent - I can't figure out a way to query specs with label selectors. This would be ideal if someone else knew the kungfu?

**Comment by klingerf** [comment]
I already mentioned this to @grampelberg, but wanted to note that our current inject implementation only modifies pod specs; it makes no modifications to parent resources. If we want to start modifying parent resources to include the `linkerd.io/injected: true` annotation, that is going to have some implications for #2342, which proposes that we switch our auto-inject implementation to watch pods instead of parent resources (at which point we'll have no ability to modify parent resources).

**Comment by olix0r** [comment]
There are 2 separate questions here:

1. Removing the data plane. This seems like it needs to be handled as part of a normal deploy that flips the inject annotation to disabled.
2. Removing the control plane once all of the above is complete. If we label all installed resources consistently, we can probably rely on something like `kubectl delete --all-namespaces all -l linkerd.io/control-plane-component` -- however, we need to be careful that all apiextensions and webhook configurations that reference linkerd components are labeled appropriately, otherwise we can leave the cluster in an odd stae.

**Comment by adleong** [comment]
I'll just add that there are shared global resources such as CRDs which do not belong to any one control plane.  These would also need a label to facilitate deleting these as well.  Something like `linkerd.io/shared-component: true` perhaps?

**Comment by ihcsim** [comment]
In #2971, I labelled all the resources with the `linkerd.io/control-plane-ns: <ns>` label. If that label isn't the right choice, we can either still use `linkerd.io/control-plane-component: <component>` or introduce a new label (like `linkerd.io/shared-component: [true|false]` as @adleong suggested).

With `linker.io/control-plane-component`, we can still use it like `kubectl delete all -l linkerd.io/control-plane-component # an IN operator`. Here are some examples of how we can label the different resources:

```
# rbac
linkerd.io/control-plane-component: rbac

# psp
linkerd.io/control-plane-component: psp

# mwc
linkerd.io/control-plane-component: proxy-injector

# crd
linkerd.io/control-plane-component: crd

# etc...
```
LMK what works best so I can get it in the next edge. Besides the label, `linkerd check` will need to be updated to use the right label.

**Comment by grampelberg** [comment]
@ihcsim it sounds like we have two uninstall use cases:

- Remove my thing that I installed (anything that isn’t a singleton)
- Remove *all* traces (including singletons)

`linkerd.io/control-plane-ns: <ns>` solves the first use case well. From a documentation perspective, `linkerd.io/shared-component: [true|false]` seems like the right way to solve the second use case.

I like the idea of adding `linkerd.io/control-plane-component` to everything, but I don’t have a solid use case for it right now.

@adleong WDYT?

**Comment by adleong** [comment]
That sounds right to me.  And I also am not clear on what `linkerd.io/control-plane-component` is for.  Especially given that the kind of a resource is already available in the resource metadata.

**Comment by siggy** [comment]
The `linkerd.io/control-plane-component` label is (at least) used by Prometheus to identity control plane components to scrape:
https://github.com/linkerd/linkerd2/blob/master/chart/templates/prometheus.yaml#L50

**Comment by ihcsim** [comment]
@siggy Thanks for the info. In that case, it's probably best to leave `linkerd.io/control-plane-component` alone.

**Comment by ihcsim** [comment]
https://github.com/linkerd/linkerd2/pull/2971 and https://github.com/linkerd/linkerd2/pull/3026 labelled all Linkerd resources with the `linkerd.io/control-plane-ns` label. We can now use `kubectl` to select all the control plane resources by this label:
```
kubectl -n linkerd get all,configmap,secret,crd,psp,clusterrole,clusterrolebinding,role,rolebinding,mutatingwebhookconfiguration,validatingwebhookconfiguration -l linkerd.io/control-plane-ns
```

For the data plane, we can select all injected pods with:
```
kubectl get po --all-namespaces -l linkerd.io/control-plane-ns=linkerd
```
Note that, however, this label is injected into the pod, so we can't  use it to select the parent workloads. 

To un-inject the data plane, we will have to update our `uninject` implementation to annotate the workload with `linkerd.io/inject: disabled`, and remove any existing `linkerd.io/inject: enabled` annotation. Then we can do something like:
```
kubectl get deploy --all-namespaces -l linkerd.io/inject=enabled | linkerd uninject -
```
